core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -5827,7 +5827,16 @@ object ZIO extends ZIOCompanionPlatformSpecific with ZIOCompanionVersionSpecific
     failureK: Cause[E1] => ZIO[R, E2, A2]
   ) extends Continuation
       with ZIO[R, E2, A2]
-  private[zio] final case class Sync[A](trace: Trace, eval: () => A) extends ZIO[Any, Nothing, A]
+  private[zio] abstract class Sync[A](val trace: Trace) extends ZIO[Any, Nothing, A] {
+    def eval(): A
+    def canEqual(that: Any) = that.isInstanceOf[Sync[_]]
+    def productArity = 2
+    def productElement(n: Int) = 
+      n match {
+        case 0 => trace
+        case 1 => () => eval()
+      }
```

Comment:
I tried to reproduce the behavior of `case class`. I wonder if there's a better way to do this.

Reply:
We can preserve compatibility, I think, by having a constructor and deconstructor with the same signature as the old `Sync`. Then we can merge without updating `zio-profiling`. @kyri-petrou @fwbrasil How does that sound?

Reply:
In fact, I think you can leave it as a `case class`, and just introduce `def eval(): A = null.asInstanceOf[A]` and override it in every subclass (yes, you can subclass case classes!), as well as override the stuff you already overrode.